tinygraph/tinygraph.c

```diff
@@ -506,3 +507,334 @@ bool tinygraph_reorder(
+    tinygraph_bitset_destruct(seen);
+    free(out);
+
+    return NULL;
+  }
+
+  tinygraph_array_s path = tinygraph_array_construct(0);
+
+  if (!path) {
+    free(parent);
+    free(dist);
+    tinygraph_heap_destruct(heap);
+    tinygraph_bitset_destruct(seen);
+    free(out);
+
+    return NULL;
+  }
+
+  *out = (tinygraph_dijkstra){
+    .s = UINT32_MAX,
+    .t = UINT32_MAX,
+    .dist = dist,
+    .parent = parent,
+    .path = path,
+    .weight = weights,
+    .graph = graph,
+    .seen = seen,
+    .heap = heap,
+  };
+
+  // Resets the internal state e.g. sets
+  // self-loops for the parents array
+  tinygraph_dijkstra_clear(out);
+
+  return out;
+}
+
+
+void tinygraph_dijkstra_destruct(tinygraph_dijkstra * const ctx) {
+  if (!ctx) {
+    return;
+  }
+
+  ctx->s = UINT32_MAX;
+  ctx->t = UINT32_MAX;
+
+  free(ctx->dist);
+  free(ctx->parent);
+
+  tinygraph_array_destruct(ctx->path);
+  tinygraph_heap_destruct(ctx->heap);
+  tinygraph_bitset_destruct(ctx->seen);
+
+  free(ctx);
+}
+
+
+static inline uint32_t tinygraph_saturated_add_u32(uint32_t a, uint32_t b) {
+  const uint32_t sum = a + b;
+
+  if (sum < a) {
+    return UINT32_MAX;
+  }
+
+  if (sum < b) {
+    return UINT32_MAX;
+  }
+
+  return sum;
+}
+
+
+bool tinygraph_dijkstra_shortest_path(
+    tinygraph_dijkstra_s ctx,
+    uint32_t s,
+    uint32_t t
+) {
+  TINYGRAPH_ASSERT(ctx);
+  TINYGRAPH_ASSERT(tinygraph_has_node(ctx->graph, s));
+  TINYGRAPH_ASSERT(tinygraph_has_node(ctx->graph, t));
+  TINYGRAPH_ASSERT(ctx->heap);
+  TINYGRAPH_ASSERT(ctx->seen);
+  TINYGRAPH_ASSERT(ctx->dist);
+  TINYGRAPH_ASSERT(ctx->parent);
+  TINYGRAPH_ASSERT(ctx->weight);
+
+  if (s == t) {
+    ctx->s = s;
+    ctx->t = t;
+
+    return true;
+  }
+
+  // Invalidate the retrieved path cache
+  // when source or target nodes change
+  if (s != ctx->s || t != ctx->t) {
+    tinygraph_array_clear(ctx->path);
+  }
+
+  // If the source node stays the same across searches
+  // and only the target node changes, we can re-use
+  // the internal state. Otherwise clear it and start
+  // the search from scratch with the new source node.
+  if (s != ctx->s) {
+    tinygraph_dijkstra_clear(ctx);
+
+    ctx->s = s;
+    ctx->t = t;
+    ctx->dist[s] = 0;
+
+    if (!tinygraph_heap_push(ctx->heap, s, 0)) {
+      tinygraph_dijkstra_clear(ctx);
+      tinygraph_array_clear(ctx->path);
+      return false;
+    }
+  } else {
+    // The source node is the same and we have explored t already
+    // in a previous search, this means we're done here
+    ctx->t = t;
+    ctx->dist[s] = 0;
+
+    if (tinygraph_bitset_get_at(ctx->seen, t)) {
+      return true;
+    }
+  }
+
+  while (!tinygraph_heap_is_empty(ctx->heap)) {
+    const uint32_t u = tinygraph_heap_pop(ctx->heap);
+
+    if (tinygraph_bitset_get_at(ctx->seen, u)) {
+      continue;
+    } else {
+      tinygraph_bitset_set_at(ctx->seen, u);
+    }
+
+    // The following is a bit different to the classical Dijkstra
```

Comment:
Well, to be pedantic: This because you mark a node as settled (or here `seen`) before you have finished processing it. The easier logic change is to move the `tinygraph_bitset_set_at(ctx->seen, u);` to the end of the function and terminate here if u == t.

Reply:
Yes and I believe it requires a bit more care when restarting: If we terminate early here we pop'ed `u` from the heap. Now if the user runs a subsequent query with the same `s` but a different `t` we didn't settle `u` and we'd have to restart accordingly pushing it into the heap etc.

Maybe it's a better trade-off, tho, to have potential subsequent queries with the same `s` pay the penalty of taking care of restarting. Not sure if it's worth now changing, happy to accept a pull request if you want to switch this over.